File: 2018/18/lumber.py

CLEAR = 0
LUMBERYARD = 1
TREES = 2

def main():
  with open('input.txt', 'r') as f:
    lines = [line.strip() for line in f.readlines()]
  map = {}
  for y, line in enumerate(lines):
    for x, state in enumerate(line):
      if state == '#':
        map[(x, y)] = LUMBERYARD
      elif state == '|':
        map[(x, y)] = TREES
      elif state == '.':
        map[(x, y)] = CLEAR
      else:
        raise Exception('bad state: %s' % state)
  # Part 1
  final_map = evolve(map, 1000000000)
  print show(final_map)
  print len([m for m in final_map if final_map[m] == LUMBERYARD]) * len([m for m in final_map if
    final_map[m] == TREES])

def evolve(map, steps):
  seen = {}
  s = 0
  while s < steps:
    new_map = {}
    for m in map:
      if map[m] == CLEAR:
        new_map[m] = [CLEAR, TREES][neighbor_count(map, m, TREES) >= 3]
      elif map[m] == TREES:
        new_map[m] = [TREES, LUMBERYARD][neighbor_count(map, m, LUMBERYARD) >= 3]
      elif map[m] == LUMBERYARD:
        new_map[m] = [CLEAR, LUMBERYARD][neighbor_count(map, m, LUMBERYARD) >= 1 and
            neighbor_count(map, m, TREES) >= 1]
      else:
        raise 'bad map state: %d' % map[m]
    map = new_map
    string = show(map)
    if string in seen:
      cycle_length = s - seen[string]
      print 'cycle length %d' % cycle_length
      skip_cycles = (steps - s) / cycle_length
      s += skip_cycles * cycle_length
    else:
      seen[string] = s
    s += 1
  return map

def neighbor_count(map, coord, value):
  return sum(1 for n in neighbors(coord) if map.get(n, None) == value)

def neighbors(coord):
  return [(coord[0], coord[1] + 1),
          (coord[0], coord[1] - 1),
          (coord[0] + 1, coord[1]),
          (coord[0] - 1, coord[1]),
          (coord[0] - 1, coord[1] - 1),
          (coord[0] - 1, coord[1] + 1),
          (coord[0] + 1, coord[1] - 1),
          (coord[0] + 1, coord[1] + 1)]

def show(map):
  result = ''
  max_x = max(m[0] for m in map)
  max_y = max(m[1] for m in map)
  for y in xrange(max_y+1):
    for x in xrange(max_x+1):
      if map[(x, y)] == LUMBERYARD:
        result += '#'
      elif map[(x, y)] == TREES:
        result += '|'
      elif map[(x, y)] == CLEAR:
        result += '.'
      else:
        raise Exception('bad state: %d' % map[(x, y)])
    result += '\n'
  return result

if __name__ == '__main__':
  main()

File: 2018/18/Lumber.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.util.List;
import java.util.stream.Collectors;

public final class Lumber {
  public enum State {
    CLEAR,
    LUMBERYARD,
    TREES,
  }

  public static void main(String[] args) throws Exception {
    List<String> lines = new BufferedReader(new FileReader(new File("input.txt"))).lines()
        .collect(Collectors.toList());
    int N = lines.size();

    State[][] map = new State[N][N];
    for (int y = 0; y < N; y++) {
      for (int x = 0; x < N; x++) {
        switch (lines.get(y).charAt(x)) {
          case '.':
            map[x][y] = State.CLEAR;
            break;
          case '|':
            map[x][y] = State.TREES;
            break;
          case '#':
            map[x][y] = State.LUMBERYARD;
            break;
          default:
            throw new RuntimeException("bad state");
        }
      }
    }

    //State[][] finalMap = evolve(map, 10);
    State[][] finalMap = evolve(map, 1000000000);
    show(finalMap);
    System.out.println(totalCount(finalMap, State.TREES) * totalCount(finalMap, State.LUMBERYARD));
  }

  private static State[][] evolve(State[][] map, long steps) {
    int N = map.length;
    State[][] newMap = new State[N][N];
    for (long i = 0; i < steps; i++) {
      if (i % 10000 == 0) {
        System.out.println(String.format("progress = %f", i * 1.0 / steps));
      }
      for (int x = 0; x < N; x++) {
        for (int y = 0; y < N; y++) {
          switch(map[x][y]) {
            case CLEAR:
              newMap[x][y] = neighborCount(map, x, y, State.TREES) >= 3 ? State.TREES : State.CLEAR;
              break;
            case TREES:
              newMap[x][y] = neighborCount(map, x, y, State.LUMBERYARD) >= 3 ? State.LUMBERYARD : State.TREES;
              break;
            case LUMBERYARD:
              newMap[x][y] = neighborCount(map, x, y, State.LUMBERYARD) >= 1
                && neighborCount(map, x, y, State.TREES) >= 1 ? State.LUMBERYARD : State.CLEAR;
              break;
            default:
              throw new RuntimeException("bad state");
          }
        }
      }
      // Swap and re-use previous array
      State[][] t = map;
      map = newMap;
      newMap = t;
    }
    return map;
  }

  private static int[][] neighbors(int x, int y) {
    return new int[][]{{x, y+1},
                       {x, y-1},
                       {x+1, y},
                       {x-1, y},
                       {x+1, y+1},
                       {x+1, y-1},
                       {x-1, y+1},
                       {x-1, y-1}};
  }

  private static int neighborCount(State[][] map, int x, int y, State state) {
    int count = 0;
    int N = map.length;
    for (int[] coord : neighbors(x, y)) {
      if (coord[0] >= 0 && coord[0] < N && coord[1] >= 0 && coord[1] < N && map[coord[0]][coord[1]] == state) {
        count += 1;
      }
    }
    return count;
  }

  private static int totalCount(State[][] map, State state) {
    int N = map.length;
    int count = 0;
    for (int x = 0; x < N; x++) {
      for (int y = 0; y < N; y++) {
        if (map[x][y] == state) {
          count += 1;
        }
      }
    }
    return count;
  }

  private static void show(State[][] map) {
    int N = map.length;

    for (int y = 0; y < N; y++) {
      for (int x = 0; x < N; x++) {
        switch (map[x][y]) {
          case CLEAR:
            System.out.print('.');
            break;
          case TREES:
            System.out.print('|');
            break;
          case LUMBERYARD:
            System.out.print('#');
            break;
          default:
            throw new RuntimeException("bad state");
        }
      }
      System.out.println();
    }
  }
}
